streamer/src/sendmmsg.rs

```diff
@@ -4,157 +4,155 @@ use std::io;
 use std::net::{SocketAddr, UdpSocket};
 
 #[cfg(not(target_os = "linux"))]
-pub fn send_mmsg(sock: &UdpSocket, packets: &[(&Vec<u8>, &SocketAddr)]) -> io::Result<usize> {
-    let count = packets.len();
+pub fn batch_send(
+    sock: &UdpSocket,
+    packets: &[(&[u8], &SocketAddr)],
+) -> Result<(), (io::Error, usize)> {
+    let mut total_sent = 0;
+    let mut erropt = None;
     for (p, a) in packets {
-        sock.send_to(p, *a)?;
+        if let Err(e) = sock.send_to(p, *a) {
+            if erropt.is_none() {
+                erropt = Some(e);
+            }
+        } else {
+            total_sent += 1;
+        }
     }
 
-    Ok(count)
+    if let Some(err) = erropt {
+        Err((err, total_sent))
+    } else {
+        Ok(())
+    }
 }
 
 #[cfg(target_os = "linux")]
-use libc::{iovec, mmsghdr, sockaddr_in, sockaddr_in6};
+use libc::{iovec, mmsghdr, sockaddr_in, sockaddr_in6, sockaddr_storage};
 
 #[cfg(target_os = "linux")]
 fn mmsghdr_for_packet(
     packet: &[u8],
     dest: &SocketAddr,
     index: usize,
-    addr_in_len: u32,
-    addr_in6_len: u32,
     iovs: &mut Vec<iovec>,
-    addr_in: &mut Vec<sockaddr_in>,
-    addr_in6: &mut Vec<sockaddr_in6>,
-) -> mmsghdr {
+    addrs: &mut Vec<sockaddr_storage>,
+    hdrs: &mut Vec<mmsghdr>,
+) {
     use libc::c_void;
     use nix::sys::socket::InetAddr;
     use std::mem;
 
+    const SIZE_OF_SOCKADDR_IN: usize = mem::size_of::<sockaddr_in>();
+    const SIZE_OF_SOCKADDR_IN6: usize = mem::size_of::<sockaddr_in6>();
+
     iovs.push(iovec {
         iov_base: packet.as_ptr() as *mut c_void,
         iov_len: packet.len(),
     });
 
-    let mut hdr: mmsghdr = unsafe { mem::zeroed() };
-    hdr.msg_hdr.msg_iov = &mut iovs[index];
-    hdr.msg_hdr.msg_iovlen = 1;
-    hdr.msg_len = packet.len() as u32;
+    let hdr: mmsghdr = unsafe { mem::zeroed() };
+    hdrs.push(hdr);
+
+    let addr_storage: sockaddr_storage = unsafe { mem::zeroed() };
+    addrs.push(addr_storage);
+
+    assert!(index < hdrs.len());
```

Comment:
`debug_assert!` is enough since this is a private function.
May also add a few more sanity checks _before_ any of the `.push` lines:
```
debug_assert!(hdrs.capacity() >= index);
debug_assert!(addrs.capacity() >= index);
debug_assert!(iovs.capacity() >= index);
debug_assert_eq!(hdrs.len(), addrs.len());
debug_assert_eq!(hdrs.len(), iovs.len());
```

Reply:
done